### PR TITLE
Feature/message cherry pick

### PR DESCRIPTION
**Pull request by m1a2st**
## Why need this change? / Root cause: 
- 可以把message 移動到thread
## Changes made:
- 增加/message cherry-pick指令
- 增加/message cherry-pick to指令
## Test Scope / Change impact:
- no test scope

#130

### CODE REVIEW

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/CherryPickManager.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package tw.waterballsa.utopia
+
+import tw.waterballsa.utopia.domain.User
+
+class CherryPickManager {
+
+    private var users = mutableListOf<User>()
```
Comment (frankvicky) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/CherryPickManager.kt:7]: `val`

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/CherryPickManager.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    fun register(user: User) {
+        users.add(user)
+    }
+
+    fun unregister(userId: String) {
+        users.removeIf { it.id.value == userId }
+    }
+
+    fun contains(userId: String): Boolean {
+        return users.any { it.id.value == userId }
```
Comment (frankvicky) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/CherryPickManager.kt:18]: `fun contains(userId: String): Boolean = users.any { it.id.value == userId }`

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/CherryPickManager.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    fun getUser(userId: String): User? {
+        return users.find { it.id.value == userId }
+    }
```
Comment (frankvicky) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/CherryPickManager.kt:23]: `fun getUser(userId: String): User? = users.find { it.id.value == userId }`

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                val selectedRange = user.selectedRange
+
+                replyEphemeralMessage("start cherry-pick to $channelName")
+                cherryPickManager.unregister(userId)
+
+                channel.iterableHistory
+                    .filter {
+                        it.inSelectedRange(selectedRange) &&
+                                it.hasCherryPickReaction() &&
+                                it.isContentNotBlank()
```
Comment (frankvicky) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt:99]: 封裝成一個方法

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                    }
+                    .sortedBy { it.timeCreated }
+                    .forEach { thread!!.sendMessageEmbeds(it.createEmbedMessage()).queue() }
+            }
+        }
+    }
+
+    private fun SlashCommandInteractionEvent.replyEphemeralMessage(message: String) {
```
Comment (frankvicky) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt:107]: 改成 `replyEphemerally`

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/domain/SelectedRange.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    val endTime: LocalDateTime
+) {
+    constructor(startTime: String, endTime: String) : this(
+        startTime = startTime.addDayBeginTime().toLocalDateTime(),
+        endTime = endTime.addDayEndTime().toLocalDateTime()
+    )
+
+    fun isStartTimeAfterEndTime(): Boolean {
+        return startTime.isAfter(endTime)
+    }
```
Comment (frankvicky) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/domain/SelectedRange.kt:19]: `fun isStartTimeAfterEndTime(): Boolean = startTime.isAfter(endTime)`

**Review by frankvicky — CHANGES_REQUESTED** [review]

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
+        with(event) {
+            if (fullCommandName != "$MESSAGE $CHERRY_PICK" &&
+                fullCommandName != "$MESSAGE $CHERRY_PICK_TO"
+            ) {
+                return
+            }
+
+            // check sub command is cherry-pick
+            if (subcommandName == CHERRY_PICK) {
```
Comment (Wally5077) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt:61]: `when`

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                val userId = event.user.id
+                if (cherryPickManager.contains(userId)) {
+                    replyEphemeralMessage("you are in use")
+                    return
+                }
+                cherryPickManager.register(User(User.Id(userId), selectedRange))
+                replyEphemeralMessage("starting cherry-pick, and The Range is $startTime between $endTime, using $MARK_EMOJI to mark message which uou want to cherry pick")
+            }
+
+            if (subcommandName == CHERRY_PICK_TO) {
```
Comment (Wally5077) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt:79]: ditto `when`

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            replyEphemerally("start time must be before end time")
+            return
+        }
+
+        if (cherryPickManager.containsUser(userId)) {
+            replyEphemerally("you are using cherry-pick")
+            return
+        }
+
+        cherryPickManager.register(User(userId), dateTimeRange)
```
Comment (Wally5077) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt:89]: **CherryPick**
```java
data class CherryPick(val cherryPickerId: String, val dateTimeRange: DateTimeRange)
```

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+            }
+        }
+    }
+
+    private fun SlashCommandInteractionEvent.cherryPick() {
+        val userId = user.id
+        val startDate = getValidateDateString(START_TIME) ?: return
+        val endDate = getValidateDateString(END_TIME)
+        val startDateTime = startDate.toTaiwanEpochMilli(0)
+        val endDateTime = endDate?.toTaiwanEpochMilli(0) ?: startDate.toTaiwanEpochMilli(1)
```
Comment (Wally5077) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt:77]: 這裡還是使用 **OffsetDateTime**

File: message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatter
+
+private const val HYPHEN_DATE_PATTERN = "yyyy-MM-dd"
+
+@Component
+class MessageCherryPickListener : UtopiaListener() {
+
+    private val cherryPickManager = CherryPickManager()
```
Comment (Wally5077) [inline message-cherry-pick/src/main/kotlin/tw/waterballsa/utopia/MessageCherryPickListener.kt:32]: **CherryPickRepository**

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Review by Wally5077 — DISMISSED** [review]
**LGTM**

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Review by frankvicky — APPROVED** [review]

**Pull request merged**